src/main/java/org/moltimate/moltimatebackend/service/DockingService.java

```diff
@@ -2,13 +2,208 @@
 
 import lombok.extern.slf4j.Slf4j;
 import org.moltimate.moltimatebackend.dto.request.DockingRequest;
-import org.moltimate.moltimatebackend.model.ligand.PDBQT;
+import org.moltimate.moltimatebackend.dto.request.ExportLigand;
+import org.moltimate.moltimatebackend.dto.request.ExportRequest;
+import org.springframework.core.io.ByteArrayResource;
+import org.springframework.core.io.Resource;
+import org.moltimate.moltimatebackend.exception.DockingJobFailedException;
+import org.moltimate.moltimatebackend.exception.JobProcessingExeption;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.http.*;
 import org.springframework.stereotype.Service;
+import org.springframework.util.LinkedMultiValueMap;
+import org.springframework.util.MultiValueMap;
+import org.springframework.web.client.HttpServerErrorException;
+import org.springframework.web.client.RestTemplate;
+import org.springframework.web.multipart.MultipartFile;
+
+import java.io.*;
+import java.util.Objects;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipInputStream;
 
 @Service
 @Slf4j
 public class DockingService {
-	public PDBQT dockLigand(DockingRequest request) {
-		return new PDBQT();
+	@Value( "${autodock.url}" )
+	private String autoDockURL;
+	@Value( "${openbabel.url}" )
+	private String openBabelURL;
+
+	public String dockLigand(DockingRequest request) throws IOException {
+
+		HttpHeaders headers = new HttpHeaders();
+		headers.setContentType(MediaType.MULTIPART_FORM_DATA);
+
+		RestTemplate template = new RestTemplate();
+
+		// Begin by converting macromolecule and ligand to pdbqt format using OpenBabel.
+
+		MultiValueMap<String, Object> openBabelParams = new LinkedMultiValueMap<>();
+		openBabelParams.add("molecule_1", new ByteArrayResource( request.getMacromolecule().getBytes() ){
+			@Override
+			public String getFilename() {
+				return request.getMacromolecule().getOriginalFilename();
+			}
+		});
+		openBabelParams.add("molecule_2", new ByteArrayResource( request.getLigand().getBytes() ){
+			@Override
+			public String getFilename() {
+				return request.getLigand().getOriginalFilename();
+			}
+		});
+
+		HttpEntity<MultiValueMap<String, Object>> entityBabelPost = new HttpEntity<>(openBabelParams, headers);
+		String babelHash = template.postForEntity( openBabelURL, entityBabelPost, String.class ).getBody();
+
+		// Check the status of the file conversion every 20 seconds.
+
+		try {
+			ResponseEntity<byte[]> babelConversion =
+					template.getForEntity( openBabelURL + "?storage_hash=" + babelHash, byte[].class );
+			while( babelConversion.getStatusCode().equals(HttpStatus.MULTIPLE_CHOICES) ) {
+				try {
+					Thread.sleep( 20000 );
```

Comment:
Possibly shorten this waiting time? @etiano417

Reply:
In my testing, it normally took at least 30 seconds to convert the file, sometimes up to several minutes, so I think that 20 seconds is a good amount to not spam the openbabel server with requests.

Reply:
@tydus101 @kdm5498 the current behavior sounds good to me then. For something like this where its a usability vs performance tradeoff and where we don't have solid numbers on the impact on either in actual use, or what typical user behavior is (say, how may of these can we expect to be converted at a time on average) i'd say whatever your first instinct here is fine. Maybe down the road we can discuss calibrating it, but I wouldn't worry too much about it at this point.